README.md

```diff
@@ -2,11 +2,165 @@
 
-MIT
+# How To Use
+
+Install from npm:
+
+```
+npm install --save-dev @ason/assembly
+```
+
+Modify your asconfig to include the transform:
+
+```ts
+{
+  "options": {
+    ... // other options here
+    "transform": ["@ason/transform"]
+  }
+}
+```
+
+Import the library and serialize away!
+
+```ts
+import { ASON } from "@ason/assembly";
+
+// serialize can determine type information
+let buffer: StaticArray<u8> = ASON.serialize([3.14, 99, 25.624] as Array<f64>);
+
+// deserialize must have the type passed (to perform type assertions)
+let result: Array<f64> = ASON.deserialize<Array<f64>>(buffer);
+
+assert(result.length == 3);
+assert(result[0] == <f64>3.14);
+assert(result[1] == <f64>99);
+assert(result[2] == <f64>25.624);
+```
+
+It's also possible to save heap allocations and re-use a `Serializer` and `Deserializer` object when serializing multiple objects of the same type.
+
+```ts
+import { Serializer, Deserializer } from "@ason/assembly";
+class Vec3 {
+  constructor(public x: f32, public y: f32, public z: f32) {}
+}
+
+let result = new Array<StaticArray<u8>>(); // an array of buffers
+let serializer = new ASON.Serializer<Vec3>();
+
+for (let i = 0; i < 10; i++) {
+  result.push(serializer.serialize(new Vec3(1, 2, 3)));
+}
+
+let deserializer = new ASON.Deserializer<Vec3>();
+
+for (let i = 0; i < 10; i++) {
+  let vec = deserializer.deserialize(result[i]);
+  assert(vec); // make sure the reference isn't null
+  assert(vec.x == 1); // check the properties
+  assert(vec.y == 2);
+  assert(vec.x == 3);
+}
+```
+
+# Uses
+
+This library is perfect for transferring references from one module of the same type to another module of the exact same type.
+
+These serialization methods are also great for helping store references like configuration files on disk.
+
+If JSON is too verbose or requires too much space in memory, ASON is a better alternative, because it reduces overhead byte storage a very large amount.
+
+# Caveats
+
+If the modules using this library are different, then runtime type information might not match. This will result in runtime errors, `instanceof` checks failing, and undefined behavior. ASON also performs type information validation for objects at the top level, so providing the wrong reference type parameter to `ASON` will result in a runtime error.
+
+# Implementation
+
+The ASON serialization and deserialization methods use a collection of tables to describe the shape of a given reference by recording entries for every field, every reference, every array, and every data segment (string, array of data, and static arrays.) It also uses a few tables to describe how objects are linked to each other to contend with the garbage collection algorithm and assert that objects will not be freed or mishandled at deserialization time.
```

Comment:
We use a custom `Table` class, which is a glorified c-like array of unmanaged entries. It's annotated with header information about every object in the object graph.